File: src/codegen/templates/abstracttemplate.py
#
# Copyright (c) 2012 Red Hat, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#           http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import os
from codegen.utils.fileutils import FileUtils
from codegen.templates.itemplate import ITemplate
import abc


class AbstractTemplate(ITemplate):
    '''
    Providing template abstraction services
    '''

    __metaclass__ = abc.ABCMeta

    def __init__(self):
        '''
        @param noCopyrightTemplate
                   true/false (use this param when no need to fetch CopyrightTemplate)
        '''
        self.name = self.__getTemplateName()
        self.path = self.__getTemplatePath()
        self.template = self._loadTemplate();

    def __initResources(self):
        self.name = None
        self.template = None
        self.copyrightTemplate = None

    def __getTemplateName(self):
        '''
        Retrives template name
        '''

        return self.__class__.__name__

    def __getTemplatePath(self):
        '''
        Builds template path
        '''

        pathArr = __file__.split(os.path.sep)
        pathArr[-1] = self.name.lower()
        return os.path.sep.join(pathArr)


    def _loadTemplate(self):
        '''
        Loads template in given context

        @return template
        '''
        ITemplate._loadTemplate(self)

        try:
            return self.__readFileTemplate()
        except Exception, e:
            raise Exception("Template \"" + self.getName() + "\" not found.", e);

    def generate(self, params={}):
        '''
         Produces content from the template
        
         @param params: parameters to bind in to the template
         
         @return content
        '''

        ITemplate.generate(self, params=params)

        return self.getTemplate() % params

    def __readFileTemplate(self):
        '''
        Reads actual template file
        
        @return template content
        '''
        return FileUtils.getContent(self.path);

    def getName(self):
        '''
        @return template name
        '''

        return self.name;

    def getTemplate(self):
        '''
        @return abstract template form
        '''

        return self.template;

File: generator/src/main/java/org/ovirt/engine/sdk/generator/templates/AbstractTemplate.java
//
// Copyright (c) 2014 Red Hat, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package org.ovirt.engine.sdk.generator.templates;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public abstract class AbstractTemplate {
    /**
     * The key that surrounds variables in the template text.
     */
    private String KEY_WRAP = "$";

    /**
     * The name of the template.
     */
    private String name;

    /**
     * The text of the template.
     */
    private String template;

    /**
     * The values of the variables to substitute in the generated text.
     */
    private Map<String, Object> variables = new HashMap<>();

    public AbstractTemplate() {
        this.name = getClass().getSimpleName();
        this.template = loadTemplate();
    }

    /**
     * Loads template in given context
     * 
     * @return template
     */
    public String loadTemplate() {
        try (InputStream in = this.getClass().getResourceAsStream(name)) {
            if (in == null) {
                throw new RuntimeException("Template \"" + name + "\" not found.");
            }
            try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
                byte[] buffer = new byte[1024];
                int count;
                while ((count = in.read(buffer)) != -1) {
                    out.write(buffer, 0, count);
                }
                return new String(out.toByteArray(), "UTF-8");
            }
        }
        catch (IOException exception) {
            throw new RuntimeException("Error loading template \"" + name + "\".", exception);
        }
    }

    /**
     * Assigns a value to a variable.
     */
    public void set(String name, Object value) {
        variables.put(name, value);
    }

    /**
     * Assign values to several variables.
     */
    public void set(Map<String, String> map) {
        variables.putAll(map);
    }

    /**
     * Evaluates the template and returns the generated text.
     */
    public String evaluate() {
        String text = template;
        for (Map.Entry<String, Object> variable : variables.entrySet()) {
            String key = KEY_WRAP + variable.getKey() + KEY_WRAP;
            Object value = variable.getValue();
            if (value != null) {
                text = text.replace(key, value.toString());
            }
        }
        return text;
    }
}
